controller/src/lib.rs

```diff
@@ -20,114 +25,204 @@ pub mod plan;
-                    let cluster = state
-                        .cluster(&schedule_request.value.cluster)
-                        .ok_or_else(|| anyhow!("Expected cluster to exist."))?;
-
-                    let backend_state = cluster
-                        .backend(&backend)
-                        .ok_or_else(|| anyhow!("Lock held by a backend that doesn't exist."))?;
+async fn spawn_backend(
+    ref nats: TypedNats,
+    drone: DroneId,
+    schedule_request: &ScheduleRequest,
+) -> anyhow::Result<(ScheduleResponse, Option<u64>)> {
+    let timer = Timer::new();
+    let spawn_request = schedule_request.schedule(&drone);
+    match nats.request(&spawn_request).await {
+        Ok(true) => {
+            tracing::info!(
+                duration=?timer.duration(),
+                backend_id=%spawn_request.backend_id,
+                %drone,
+                "Drone accepted backend."
+            );
+
+            let seq_id = nats
+                .publish_jetstream(&WorldStateMessage {
+                    cluster: schedule_request.cluster.clone(),
+                    message: ClusterStateMessage::BackendMessage(BackendMessage {
+                        backend: spawn_request.backend_id.clone(),
+                        message: BackendMessageType::Assignment {
+                            drone: drone.clone(),
+                            lock: schedule_request.lock.clone(),
+                            bearer_token: spawn_request.bearer_token.clone(),
+                        },
+                    }),
+                })
+                .await?;
+
+            tracing::info!(
+            logical_time=?seq_id,
+            "backend state updated at time"
+                );
+
+            Ok((
+                ScheduleResponse::Scheduled {
+                    drone,
+                    backend_id: spawn_request.backend_id,
+                    bearer_token: spawn_request.bearer_token.clone(),
+                    spawned: true,
+                },
+                Some(seq_id),
+            ))
+        }
+        Ok(false) => {
+            tracing::warn!("Drone rejected backend.");
+            Ok((ScheduleResponse::NoDroneAvailable, None))
+        }
+        Err(error) => {
+            tracing::warn!(?error, "Scheduler returned error.");
+            Ok((ScheduleResponse::NoDroneAvailable, None))
+        }
+    }
+}
 
-                    let drone = backend_state.drone.clone().ok_or_else(|| {
-                        anyhow!("Lock held by a backend without a drone assignment.")
-                    })?;
+fn locked_backend(
+    state: &StateHandle,
+    cluster_name: &ClusterName,
+    lock: String,
+) -> anyhow::Result<BackendId> {
+    state
+        .state()
+        .cluster(cluster_name)
+        .ok_or(anyhow!("no cluster"))?
+        .locked(&lock)
+        .ok_or(anyhow!("no backend"))
+}
 
-                    let bearer_token = backend_state.bearer_token.clone();
+fn fetch_backend(
```

Comment:
I think of "fetch" as implying that a network trip or at least hitting disk is involved. Maybe `schedule_request_for_existing_backend`?